content/source/docs/extend/terraform-0.12-compatibility.html.md

```diff
@@ -0,0 +1,329 @@
+defined by the provider, never a mixture. Think of a nested resource as being
+conceptually a separate object that happens to be nested, whereas an argument
+is simply a property of the main object.
+
+## Inaccurate Plans
+
+The intended contract for Terraform's plan phase is that the provider should
+produce as accurate as possible a description of what each resource object
+will look like after the apply operation is completed. Any attribute value
+that is not set in configuration and whose default cannot be predicted until
+apply time is marked as "unknown", as a placeholder for the final value.
+
+In Terraform 0.11 and prior, Terraform Core did not enforce that the final
+result be consistent with what was planned. If a provider produced a final
+result that disagreed with any known attribute in the plan, Terraform would
+just accept it and save it, most of the time letting that inconsistency go
+unnoticed.
+
+However, along with violating user expectations this would also tend to lead
+to errors on downstream resources including the message "diffs didn't match
+during apply". When Terraform 0.11 and prior returns this error, it is saying
+that when it re-ran the resource plan during the apply phase to incorporate
+values learned so far, the new plan had attribute values that were not equal
+to what was originally planned. This is because the downstream resource plan
+was derived from a predicted result from elsewhere, but the final result
+did not match the plan and thus the new plan is different.
+
+Terraform 0.12 includes a new safety check to detect when a provider produces
+a result that is inconsistent with what was planned. The error message text
+in that case will be similar to the following:
+
+```
+Error: Provider produced inconsistent result after apply
+
+When applying changes to null_resource.example, provider "null" produced an
+unexpected new value for .triggers["foo"]: was cty.StringVal("a"), but now
+cty.StringVal("b").
+
+This is a bug in the provider, which should be reported in the provider's
+own issue tracker.
+```
+
+Because such inconsistencies turned out to be quite common in existing
+provider implementations (a result of this not being enforced before), Terraform
+0.12 does not enforce this as a hard error for providers using the current
+version of the SDK, and so such problems will for now continue to return the
+new equivalent of the "diffs didn't match during apply" message, which now
+as the following structure:
+
+```
+Error: Provider produced inconsistent final plan
+
+When expanding the plan for null_resource.downstream to include new values
+learned so far during apply, provider "null" produced an invalid new value for
+.triggers["from_other"]: was cty.StringVal("a"), but now cty.StringVal("b").
+
+This is a bug in the provider, which should be reported in the provider's own
+issue tracker.
+```
+
+This other error unfortunately is reported from the perspective of the
```

Comment:
"unfortunately" here is more confusing than helpful, because the information that's important is at the end of the sentence. If you wanted to emphasize that it's unfortunate, it would make more sense after saying "the error reports THERE rather than HERE" if you want to, but I think it's cut-able